executor/write.go

```diff
@@ -65,50 +70,50 @@ func updateRecord(ctx context.Context, h int64, oldData, newData []types.Datum,
 			}
 			t.RebaseAutoID(val, true)
 		}
-		casted, err := table.CastValue(ctx, newData[i], col.ToInfo())
+		cmp, err := newData[i].CompareDatum(sc, oldData[i])
 		if err != nil {
 			return false, errors.Trace(err)
 		}
-		newData[i] = casted
-		touched[i] = true
-		assignExists = true
-	}
-
-	// If no assign list for this table, no need to update.
-	if !assignExists {
-		return false, nil
+		if cmp != 0 {
+			changed = true
+			touched[i] = true
+			if col.IsPKHandleColumn(t.Meta()) {
+				handleChanged = true
+			}
+		} else {
+			if mysql.HasOnUpdateNowFlag(col.Flag) && touched[i] {
+				// It's for "UPDATE t SET ts = ts" and ts is a timestamp.
+				onUpdateNoChange[i] = true
```

Comment:
for this case:
```sql
update t set ts_field = ts_field;
```
the `ts_field` has attributes `timestamp not null default current_timestamp on update current_timestamp`.

Suppose the `offset` of `ts_field` is 3, so `touched[3]` is `false`, because it's not changed. But we can't set it to `current_timestamp`, because we specied that field explicitly in `set list`.